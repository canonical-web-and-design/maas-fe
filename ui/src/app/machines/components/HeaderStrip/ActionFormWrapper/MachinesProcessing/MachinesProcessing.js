import { Loader } from "@canonical/react-components";
import { useSelector } from "react-redux";
import PropTypes from "prop-types";
import React, { useEffect, useRef } from "react";

import {
  general as generalSelectors,
  machine as machineSelectors,
} from "app/base/selectors";

// Handle checking when a value has cycled from false to true.
const useCycled = (newState, onCycled) => {
  const previousState = useRef(newState);
  useEffect(() => {
    if (newState && !previousState.current) {
      onCycled();
    }
    if (previousState.current !== newState) {
      previousState.current = newState;
    }
  }, [newState, onCycled]);
};

const MachinesProcessing = ({
  action,
  hasErrors = false,
  machinesProcessing,
  setProcessing,
  setSelectedAction,
}) => {
  const processingStarted = useRef(false);
  const selectedMachines = useSelector(machineSelectors.selected);
  const selectedMachinesCount = selectedMachines.length;
  const { sentence } = useSelector((state) =>
    generalSelectors.machineActions.getByName(state, action)
  );
  const selectedSavingCount = machinesProcessing.length;
  if (processingStarted.current === false && selectedSavingCount > 0) {
    processingStarted.current = true;
  }
  const processedCount = processingStarted.current
    ? selectedMachinesCount - selectedSavingCount
    : 0;

  // If all the machines have finished processing then update the state.
  useCycled(selectedSavingCount === 0, () => {
    if (!hasErrors) {
      setProcessing(false);
      setSelectedAction(null);
    }
  });

  // If the machines are processing and there are errors then exit the
  // processing state.
  useCycled(hasErrors, () => {
    if (processingStarted) {
      setProcessing(false);
    }
  });

  return (
    <p>
      <Loader inline className="u-no-margin" />
      {processedCount} of {selectedMachinesCount} nodes are transitioning to{" "}
      {sentence.toLowerCase()}.
    </p>
  );
};

MachinesProcessing.propTypes = {
  action: PropTypes.string.isRequired,
  hasErrors: PropTypes.bool,
  machinesProcessing: PropTypes.array.isRequired,
  setProcessing: PropTypes.func.isRequired,
  setSelectedAction: PropTypes.func.isRequired,
};

export default MachinesProcessing;
