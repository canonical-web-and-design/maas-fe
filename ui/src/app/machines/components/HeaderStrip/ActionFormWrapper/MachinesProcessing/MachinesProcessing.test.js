import { MemoryRouter } from "react-router-dom";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import React from "react";

import MachinesProcessing from "./MachinesProcessing";

const mockStore = configureStore();

describe("MachinesProcessing", () => {
  let state;
  beforeEach(() => {
    state = {
      general: {
        machineActions: {
          data: [{ name: "set-pool", sentence: "change those pools" }],
        },
      },
      machine: {
        items: [],
        selected: [1, 2],
      },
    };
  });

  it("can render", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <MachinesProcessing
            action="set-pool"
            machinesProcessing={[{ id: 1 }, { id: 2 }]}
            setProcessing={jest.fn()}
            setSelectedAction={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("MachinesProcessing")).toMatchSnapshot();
  });

  it("can set the processed count before it has started", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <MachinesProcessing
            action="set-pool"
            machinesProcessing={[{ id: 1 }, { id: 2 }]}
            setProcessing={jest.fn()}
            setSelectedAction={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("p").text()).toEqual(
      "0 of 2 nodes are transitioning to change those pools."
    );
  });

  it("can show the processed count", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <MachinesProcessing
            action="set-pool"
            machinesProcessing={[{ id: 2 }]}
            setProcessing={jest.fn()}
            setSelectedAction={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("p").text()).toEqual(
      "1 of 2 nodes are transitioning to change those pools."
    );
  });

  it("calls functions when processing is complete", () => {
    const setProcessing = jest.fn();
    const setSelectedAction = jest.fn();
    const store = mockStore(state);
    const Proxy = ({ machinesProcessing }) => (
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <MachinesProcessing
            action="set-pool"
            machinesProcessing={machinesProcessing}
            setProcessing={setProcessing}
            setSelectedAction={setSelectedAction}
          />
        </MemoryRouter>
      </Provider>
    );
    const wrapper = mount(<Proxy machinesProcessing={[{ id: 2 }]} />);
    expect(setProcessing).not.toHaveBeenCalled();
    expect(setSelectedAction).not.toHaveBeenCalled();
    wrapper.setProps({ machinesProcessing: [] });
    expect(setProcessing).toHaveBeenCalled();
    expect(setSelectedAction).toHaveBeenCalled();
  });
});
