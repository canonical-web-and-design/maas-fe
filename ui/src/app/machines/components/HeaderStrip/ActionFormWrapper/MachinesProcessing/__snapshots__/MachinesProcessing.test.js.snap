// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MachinesProcessing can render 1`] = `
<MachinesProcessing
  action="set-pool"
  machinesProcessing={
    Array [
      Object {
        "id": 1,
      },
      Object {
        "id": 2,
      },
    ]
  }
  setProcessing={[MockFunction]}
  setSelectedAction={[MockFunction]}
>
  <p>
    <Spinner
      className="u-no-margin"
      inline={true}
    >
      <span
        className="u-no-margin p-spinner p-text--default p-spinner--inline"
      >
        <i
          className="p-icon--spinner u-animation--spin"
        />
      </span>
    </Spinner>
    0
     of 
    2
     nodes are transitioning to
     
    change those pools
    .
  </p>
</MachinesProcessing>
`;
